src/core/server/config/env.ts

```diff
@@ -58,43 +57,63 @@ export class Env {
   public readonly staticFilesDir: string;
 
   /**
-   * @internal
+   * Information about Kibana package (version, build number etc.).
    */
-  constructor(readonly homeDir: string, private readonly options: EnvOptions) {
-    this.configDir = resolve(this.homeDir, 'config');
-    this.corePluginsDir = resolve(this.homeDir, 'core_plugins');
-    this.binDir = resolve(this.homeDir, 'bin');
-    this.logDir = resolve(this.homeDir, 'log');
-    this.staticFilesDir = resolve(this.homeDir, 'ui');
-  }
+  public readonly packageInfo: Readonly<PackageInfo>;
 
-  public getConfigFile() {
-    const defaultConfigFile = this.getDefaultConfigFile();
-    return this.options.config === undefined ? defaultConfigFile : this.options.config;
-  }
+  /**
+   * Mode Kibana currently run in (development or production).
+   */
+  public readonly mode: Readonly<EnvironmentMode>;
 
   /**
    * @internal
    */
-  public getLegacyKbnServer(): LegacyKbnServer | undefined {
-    return this.options.kbnServer;
-  }
+  public readonly legacy: EventEmitter;
 
   /**
-   * Gets information about Kibana package (version, build number etc.).
+   * Arguments provided through command line.
    */
-  public getPackageInfo() {
-    return this.options.packageInfo;
-  }
+  public readonly cliArgs: Readonly<Record<string, any>>;
 
   /**
-   * Gets mode Kibana currently run in (development or production).
+   * Paths to the configuration files.
    */
-  public getMode() {
-    return this.options.mode;
-  }
+  public readonly configs: ReadonlyArray<string>;
+
+  /**
+   * Indicates that this Kibana instance is run as development Node Cluster master.
+   */
+  public readonly isDevClusterMaster: boolean;
+
+  /**
+   * @internal
+   */
+  constructor(readonly homeDir: string, options: EnvOptions) {
+    this.configDir = resolve(this.homeDir, 'config');
+    this.corePluginsDir = resolve(this.homeDir, 'core_plugins');
+    this.binDir = resolve(this.homeDir, 'bin');
+    this.logDir = resolve(this.homeDir, 'log');
+    this.staticFilesDir = resolve(this.homeDir, 'ui');
+
+    this.cliArgs = Object.freeze(options.cliArgs);
+    this.configs = Object.freeze(options.configs);
+    this.isDevClusterMaster = options.isDevClusterMaster;
+
+    this.mode = Object.freeze<EnvironmentMode>({
+      dev: this.cliArgs.dev,
+      name: this.cliArgs.dev ? 'development' : 'production',
+      prod: !this.cliArgs.dev,
+    });
+
+    const isKibanaDistributable = pkg.build && pkg.build.distributable === true;
+    this.packageInfo = Object.freeze({
+      branch: pkg.branch,
+      buildNum: isKibanaDistributable ? pkg.build.number : Number.MAX_SAFE_INTEGER,
+      buildSha: isKibanaDistributable ? pkg.build.sha : 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
+      version: pkg.version,
+    });
 
-  private getDefaultConfigFile() {
-    return resolve(this.configDir, 'kibana.yml');
+    this.legacy = new EventEmitter();
```

Comment:
Yes, my plan is to get rid of it in https://github.com/elastic/kibana/pull/22190. I still don't like any approach to be honest since I don't want to pass internal server reference around and be a part of any public contract :/ But there is no other choice for now.